crates/masonry/src/widget/widget_mut.rs

```diff
@@ -29,14 +30,48 @@ use crate::{Widget, WidgetId, WidgetState};
 ///
 /// See [`declare_widget`](crate::declare_widget) for details.
 pub struct WidgetMut<'a, W: Widget + StoreInWidgetMut> {
-    pub(crate) parent_widget_state: &'a mut WidgetState,
-    pub(crate) inner: W::Mut<'a>,
+    pub(crate) parent_widget_state: ManuallyDrop<&'a mut WidgetState>,
+    pub(crate) inner: ManuallyDrop<W::Mut<'a>>,
+    /// Whether this `WidgetMut` is the "root" `WidgetMut` of the reborrow tree, i.e.
+    /// whether we need to call merge_up in drop, or whether that will happen later
+    root: bool,
+}
+
+impl<'a, W: Widget + StoreInWidgetMut> WidgetMut<'a, W> {
+    pub(crate) fn new(parent_widget_state: &'a mut WidgetState, inner: W::Mut<'a>) -> Self {
```

Comment:
I think all of this unsafe code is only "safe" because it's only possible to construct this type within masonry isn't it?

Not super comfortable with this, because it's... well unsafe as well as (slightly) complex (not the best combination). Not sure whether all of that could be achieved with just some reference counting.